src/main/scala/cromwell/engine/WorkflowActor.scala

```diff
@@ -1,79 +1,80 @@
 import scala.language.postfixOps
-import scala.util.{Failure, Success}
 
 
 object WorkflowActor {
-
   sealed trait WorkflowActorMessage
-
+  case object Complete extends WorkflowActorMessage
   case object Start extends WorkflowActorMessage
-  case object Started extends WorkflowActorMessage
-  case class Done(outputs: Map[String, WdlValue]) extends WorkflowActorMessage
   case class Failed(failure: String) extends WorkflowActorMessage
   case object GetOutputs extends WorkflowActorMessage
 
-  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) =
+  def props(id: WorkflowId, binding: WdlBinding, coercedInputs: WorkflowCoercedInputs, backend: Backend) = {
     Props(new WorkflowActor(id, binding, coercedInputs, backend))
+  }
 
   implicit val ActorTimeout = Timeout(5 seconds)
-}
-
-/**
- * Actor to manage the execution of a single workflow.
- */
-case class WorkflowActor private(id: WorkflowId,
-                                 binding: WdlBinding,
-                                 actualInputs: WorkflowCoercedInputs,
-                                 backend: Backend) extends Actor {
 
-  private val log = Logging(context.system, this)
+  sealed trait WorkflowActorData
+  case object NoData extends WorkflowActorData // Irritating that it takes 2 lines to not use a feature
+}
 
+case class WorkflowActor(id: WorkflowId,
+                         binding: WdlBinding,
+                         actualInputs: WorkflowCoercedInputs,
+                         backend: Backend) extends LoggingFSM[WorkflowState, WorkflowActorData] {
+  implicit val timeout = Timeout(5 seconds)
   private val storeActor = context.actorOf(StoreActor.props(binding, actualInputs))
 
-  def receive: Receive = unstarted
+  startWith(WorkflowSubmitted, NoData)
 
-  /** Unstarted `Receive` state that only responds to `Start` messages. */
-  def unstarted: Receive = LoggingReceive {
-    case WorkflowActor.Start => handleInitialStart()
-    case bad => fail(bad, "unstarted")
+  when(WorkflowSubmitted) {
+    case Event(WorkflowActor.Start, _) => goto(WorkflowRunning)
   }
 
-  /** Started `Receive` state. */
-  def started: Receive = LoggingReceive {
-    case CallActor.Started(call) =>
+  when(WorkflowRunning) {
+    case Event(CallActor.Started(call), _) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Running)
+      stay()
 
-    case CallActor.Completed(completedCall, callOutputs) =>
-      handleCallCompleted(completedCall, callOutputs)
+    case Event(CallActor.Completed(completedCall, callOutputs), _) =>
+      storeActor ! CallCompleted(completedCall, callOutputs)
+      startRunnableCalls()
+      stay()
 
-    case CallActor.Failed(call, failure) =>
+    case Event(CallActor.Failed(call, failure), _) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Failed)
-      context.parent ! WorkflowActor.Failed(failure)
+      // FIXME: Losing the 'failure' here ...
+      goto(WorkflowFailed)
```

Comment:
That doesn't send to parent though. We could store the failure mode which could be queried after the fact. This is one of the FIXMEs where I wanted to have a convo on the topic before proceeding.